wasm/datatypes/instructions.py

```diff
@@ -0,0 +1,69 @@
+from collections.abc import (
+    Sequence,
+)
+from typing import (
+    TYPE_CHECKING,
+    Iterator,
+    Tuple,
+    overload,
+)
+
+if TYPE_CHECKING:
+    from wasm.instructions import (  # noqa: F401
+        BaseInstruction,
+    )
+
+
+class InstructionSequence(Sequence):
+    _instructions: Tuple['BaseInstruction', ...]
+
+    def __init__(self, instructions: Tuple['BaseInstruction', ...]) -> None:
+        self._instructions = instructions
+        self._idx = -1
+
+    def __str__(self) -> str:
+        return f"[{' > '.join((str(instr) for instr in self._instructions))}]"
+
+    def __repr__(self) -> str:
+        return f"InstructionSequence({str(self)})"
+
+    def __len__(self) -> int:
+        return len(self._instructions)
+
+    def __next__(self) -> 'BaseInstruction':
+        self._idx += 1
+        try:
+            return self._instructions[self._idx]
+        except IndexError:
+            raise StopIteration
+
+    def __iter__(self) -> Iterator['BaseInstruction']:
+        while self._idx < len(self._instructions) - 1:
+            yield next(self)
+
+    @overload
+    def __getitem__(self, idx: int) -> 'BaseInstruction':
+        pass
+
+    @overload  # noqa: 811
+    def __getitem__(self, s: slice) -> 'InstructionSequence':
+        pass
+
+    def __getitem__(self, item):  # noqa: F811
+        if isinstance(item, int):
+            return self._instructions[item]
+        elif isinstance(item, slice):
+            return type(self)(self._instructions[item])
+        else:
+            raise TypeError(f"Unsupported key type: {type(item)}")
+
+    @property
+    def pc(self) -> int:
```

Comment:
What does `pc` stand for?

Reply:
'program counter', like 'instruction counter'; an index into the `InstructionSequence`. this is how the execution engine knows "what is the next instruction to execute"

Reply:
I feel like that would make a better name for it ;)

Reply:
https://github.com/ethereum/py-evm/blob/master/eth/vm/code_stream.py#L60-L62

I just defaulted to the same name used in the EVM.  I'll add a docstring to help clarify what this value represents.